subworkflows/nf-core/bam_downsample_samtools/main.nf

```diff
@@ -0,0 +1,56 @@
+include { SAMTOOLS_DEPTH } from '../../../modules/nf-core/samtools/depth'
+include { GAWK           } from '../../../modules/nf-core/gawk'
+include { SAMTOOLS_VIEW  } from '../../../modules/nf-core/samtools/view'
+
+workflow BAM_DOWNSAMPLE_SAMTOOLS {
+
+    take:
+    ch_bam    // channel: [ [id], bam, bai ]
+    ch_depth  // channel: [ [depth], depth ]
+    ch_fasta  // channel: [ [genome], fasta, fai ]
+
+    main:
+    ch_versions      = Channel.empty()
+
+    // Compute mean depth
+    SAMTOOLS_DEPTH(ch_bam, [[], []])
+    ch_versions = ch_versions.mix(SAMTOOLS_DEPTH.out.versions.first())
+
+    // Use GAWK to get mean depth
+    GAWK(SAMTOOLS_DEPTH.out.tsv, [])
+    ch_versions = ch_versions.mix(GAWK.out.versions.first())
+
+    // Compute downsampling factor
+    ch_depth_factor = GAWK.out.output
+        .splitCsv(header: false, sep:'\t')
+        .map{ metaI, row ->
+            [ metaI, row[0] as Float ]
+        }
+        .combine(ch_depth)
+        .map{ metaI, mean, metaD, depth ->
+            [ metaI, metaI + metaD, depth as Float / mean ]
+        }
+
+    // Add all necessary channel for downsampling
+    ch_input_downsample = ch_bam
+        .combine(ch_depth_factor, by : 0)
+        .map{ metaI, bam, index, metaID, depth ->
+            [ metaID, bam, index, [], depth ]
+        }
```

Comment:
Does metaI + metaD become metaID? Since you're removing it later, can't you: 

```suggestion
            [ meta, depth as Float / mean ]
        }

    // Add all necessary channel for downsampling
    ch_input_downsample = ch_bam
        .join(ch_depth_factor)
        .map{ meta, bam, index, depth ->
            [ meta, bam, index, [], depth ]
        }
```